ToDoApp/ToDoApp/SignUp/SignUpViewModel.swift

```diff
@@ -0,0 +1,71 @@
+//
+//  SignUpViewModel.swift
+//  ToDoApp
+//
+//  Created by Trong Hieu Vuong on 2022-12-20.
+//
+
+import Foundation
+import FirebaseAuth
+enum SignUpErrors: Identifiable, Hashable {
+    var id: Int {
+        hashValue
+    }
+    case invalidEmail
+    case invalidPassword
+    case invalidName
+    case authenticationError
+}
+@MainActor
+class SignUpViewModel: ObservableObject {
+    // MARK: Internal Properties
+    @Published var error: SignUpErrors?
+    @Published var email: String = ""
+    @Published var password: String = ""
+    @Published var name: String = ""
+    @Published var isCheckBoxTapped: Bool = false
+    private let generateUser: GenerateUser = GenerateUser()
+    private func trimTextField() {
+        self.email = email.trimmingCharacters(in: .whitespacesAndNewlines)
+        self.password = password.trimmingCharacters(in: .whitespacesAndNewlines)
+        self.name = name.trimmingCharacters(in: .whitespacesAndNewlines)
+    }
+    func signUp() {
+        trimTextField()
+        if self.name == "" {
+            error = .invalidPassword
+            return
+        }
+        if self.email == "" {
+            error = .invalidEmail
+            return
+        }
+        if !isPasswordValid(self.password) {
+            error = .invalidPassword
+            return
+        }
+        Auth.auth().createUser(withEmail: self.email, password: self.password) {
```

Comment:
User creation should be implemented in a separate type the same way the TodoService and Repository are implemented. Direct usage of the Firebase in the implementation of another entity violates dependency inversion principle